### ISSUE TITLE
Problem with RPC (request/response) on Azure Service Bus

### ISSUE THREAD

**Issue by cocowalla** [Azure/amqpnetlite#261]
I'm using Azure Service Bus, and I'm trying to figure out how to have a specific client send a request and wait for a response (RPC pattern aka request/response pattern).

I think the [Interop.Client](https://github.com/Azure/amqpnetlite/tree/master/Examples/Interop/Interop.Client) and [Interop.Server](https://github.com/Azure/amqpnetlite/blob/master/Examples/Interop/Interop.Server) demonstrate this, but I'm unable to get it working with Azure Service Bus - it fails [here](https://github.com/Azure/amqpnetlite/blob/9a68ad68d4c86662d2ece366ed9f89ff00ac341a/Examples/Interop/Interop.Client/Interop.Client.cs#L72), as `a.Source` is null.

Is this the correct way to do RPC, and if so, are you aware of any changes that need to be made to work with Azure Service Bus?

Thanks!

**Comment by xinchen10** [comment]
Azure SB does not support dynamic source yet. When you create the receiver link, try setting a queue name as the address in Source.

**Comment by cocowalla** [comment]
Problem is that I will have thousands of requestors - if they all use the same queue for responses, how can I ensure that each response goes to the correct requestor?

Sorry if this is obvious, I'm coming to AMQP 1.0 from RabbitMQ, and it really is very different!

**Comment by xinchen10** [comment]
You are right. AMQP 1.0 is completely different from the early drafts. You scenario is very common and that's what the dynamic node is for, but unfortunately SB does not support it as of today.

**Comment by cocowalla** [comment]
I consider request/response to be a fundamental pattern of messaging systems, it's kind of incredible that Azure Service Bus doesn't support it! I checked to see if I could use the REST API for request/response instead... nope, that doesn't seem possible either!

Anyway, feel free to close this if there's nothing to be done here.

**Comment by DarinMacRae** [comment]
@cocowalla I'm not sure if you have specific requirements for RPC over what you get with using Sessions but that is something to look into. Using sessions to correlate a request with a response works just fine in ASB. Drop me a line if you haven't given up on ASB already.

**Comment by cocowalla** [comment]
@DarinMacRae I haven't given up _quite_ yet, just parked it for a while :)

I did manage to get a simple PoC working using sessions, but it all feels more complicated than it should be. I think I'd want to put a layer of abstraction over it were I to use it (hah, basically to make it feel more like AMQP 0.9.1!).

It would still be really good to see some real-world examples of using sessions with ASB though.

Also, I did have some concerns and questions about security. The system I'm working on is going to be multi-tenant. Assuming a single ASB namespace, the security of sessions would seem to rely only on the randomness of the session ID. Do you know if there is a maximum length for the ID, or if there is a some good way to secure things so a nefarious client couldn't accept a session meant for another tenant?

**Comment by cocowalla** [comment]
@DarinMacRae 

> I did not find anything specific on the max length of a session id. I would assume that it is something similar to the max length of a queue name but I would trial-and-error it myself to establish some upper bound.

I did some tinkering, and it seems (strangely!) the limit is **_129_** characters. There also appear to be some restrictions on allowable characters, as Base64 encoded strings are accepted, while Ascii85 is not. I've opened a Github issue on the MS docs page, so hopefully they will update it so we don't have to guess.

Anyway, 129 characters of Base64 can encode ~768 bits - if I encode the output from a cryptographically secure PRNG, that should be plenty entropy. Actually, it's even enough to combine with your tenant ID idea - something like $"{tenantGuid}-{random}" would still provide ~576 bits of entropy before considering the 128-bit tenant ID.

> Just as Microsoft\Azure has supported multi-tenant using namespaces and secret keys that is one true way to secure them.

Yes, this was the first way I though of, but there are a few issues:

1. A limit of 100 namespaces per subscription is imposed - I wouldn't be likely to hit this soon, but it would be hit eventually
2. Given the low message rate, it seems wasteful to pay for a namespace per tenant
3. It complicates the backend, as I'll need consumers for each namespace

Still, it's a valid option that I should investigate some more.

> In case it helps you or anyone else, here is a brief summary of how I am handling it....

Yes, this is exactly how I got it working, just without the `ReplyTo` since the queue name is known. TBH, I'd prefer to create a temporary queue, similar to how RPC works with RabbitMQ, but the ASB libraries and permission model aren't compatible with letting clients create queues.

It seems like I'm on the right track though - thanks for helping me think this through 👍 

Another scenario I have is the backend needing to send messages to specific clients. With RabbitMQ, I'd setup a topic exchange and use routing keys, such that each client creates and binds a queue to the exchange - and this can be secured using a combination of topic and resource rules. What might be a good model to do this with ASB?
